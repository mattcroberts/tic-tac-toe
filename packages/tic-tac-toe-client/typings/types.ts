/* tslint:disable */
import { GraphQLResolveInfo } from "graphql";

export type Resolver<Result, Parent = any, Context = any, Args = any> = (
    parent: Parent,
    args: Args,
    context: Context,
    info: GraphQLResolveInfo
) => Promise<Result> | Result;

export type SubscriptionResolver<
    Result,
    Parent = any,
    Context = any,
    Args = any
> = {
    subscribe<R = Result, P = Parent>(
        parent: P,
        args: Args,
        context: Context,
        info: GraphQLResolveInfo
    ): AsyncIterator<R | Result>;
    resolve?<R = Result, P = Parent>(
        parent: P,
        args: Args,
        context: Context,
        info: GraphQLResolveInfo
    ): R | Result | Promise<R | Result>;
};

export interface Query {
    grid: Grid;
}

export interface Grid {
    id: string;
    gridItems: GridItem[][];
    currentPlayer: Player;
    winner?: Player | null;
    isFinished: boolean;
    size: number;
}

export interface GridItem {
    id: string;
    player?: Player | null;
}

export interface Mutation {
    executeTurn: Grid;
    newGame: Grid;
}
export interface GridQueryArgs {
    id: string;
}
export interface ExecuteTurnMutationArgs {
    id: string;
    player: string;
    x: number;
    y: number;
}

export enum Player {
    NAUGHT = "NAUGHT",
    CROSS = "CROSS"
}

export namespace QueryResolvers {
    export interface Resolvers<Context = any> {
        grid?: GridResolver<Grid, any, Context>;
    }

    export type GridResolver<R = Grid, Parent = any, Context = any> = Resolver<
        R,
        Parent,
        Context,
        GridArgs
    >;
    export interface GridArgs {
        id: string;
    }
}

export namespace GridResolvers {
    export interface Resolvers<Context = any> {
        id?: IdResolver<string, any, Context>;
        gridItems?: GridItemsResolver<GridItem[][], any, Context>;
        currentPlayer?: CurrentPlayerResolver<Player, any, Context>;
        winner?: WinnerResolver<Player | null, any, Context>;
        isFinished?: IsFinishedResolver<boolean, any, Context>;
        size?: SizeResolver<number, any, Context>;
    }

    export type IdResolver<R = string, Parent = any, Context = any> = Resolver<
        R,
        Parent,
        Context
    >;
    export type GridItemsResolver<
        R = GridItem[][],
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
    export type CurrentPlayerResolver<
        R = Player,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
    export type WinnerResolver<
        R = Player | null,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
    export type IsFinishedResolver<
        R = boolean,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
    export type SizeResolver<
        R = number,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
}

export namespace GridItemResolvers {
    export interface Resolvers<Context = any> {
        id?: IdResolver<string, any, Context>;
        player?: PlayerResolver<Player | null, any, Context>;
    }

    export type IdResolver<R = string, Parent = any, Context = any> = Resolver<
        R,
        Parent,
        Context
    >;
    export type PlayerResolver<
        R = Player | null,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
}

export namespace MutationResolvers {
    export interface Resolvers<Context = any> {
        executeTurn?: ExecuteTurnResolver<Grid, any, Context>;
        newGame?: NewGameResolver<Grid, any, Context>;
    }

    export type ExecuteTurnResolver<
        R = Grid,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context, ExecuteTurnArgs>;
    export interface ExecuteTurnArgs {
        id: string;
        player: string;
        x: number;
        y: number;
    }

    export type NewGameResolver<
        R = Grid,
        Parent = any,
        Context = any
    > = Resolver<R, Parent, Context>;
}
