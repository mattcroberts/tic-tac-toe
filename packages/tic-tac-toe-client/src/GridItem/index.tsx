import * as React from "react";

import "./GridItem.css";

export interface IItemState {
    player: string;
}
interface IGridItemProps {
    colN: number;
    rowN: number;
    itemState: IItemState;
    onItemClick: (player: string, x: number, y: number) => void;
}

class GridItem extends React.PureComponent<IGridItemProps, any> {
    constructor(props: IGridItemProps) {
        super(props);
        this.onClick = this.onClick.bind(this);
    }
    public render() {
        return (
            <div className="GridItem" onClick={this.onClick}>
                {this.props.itemState.player}
            </div>
        );
    }
    private onClick() {
        this.props.onItemClick(
            this.props.itemState.player,
            this.props.colN,
            this.props.rowN
        );
    }
}

export default GridItem;
