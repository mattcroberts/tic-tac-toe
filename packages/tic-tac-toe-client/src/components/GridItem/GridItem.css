:root {
    --border-color: black;
    --cross-color: black;
    --naught-color: black;
    --finished-color: #69656a;
}

.root {
    display: flex;
    flex-flow: column wrap;
    justify-content: center;
    flex: 1;
    border: 1px solid var(--border-color);
    font-size: 150px;
    height: 29vh;
}

.finished {
    border-color: var(--finished-color);
}

.cross {
    color: var(--cross-color);
}

.cross.finished {
    color: var(--finished-color);
}

.naught {
    color: var(--naught-color);
}

.naught.finished {
    color: var(--finished-color);
}

.top {
    border-top: none;
}
.bottom {
    border-bottom: none;
}
.left {
    border-left: none;
}
.right {
    border-right: none;
}
