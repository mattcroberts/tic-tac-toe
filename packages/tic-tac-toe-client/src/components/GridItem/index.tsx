import classnames from "classnames";
import * as React from "react";

import {
    GridItem as IGridItem,
    Symbol as ISymbol
} from "../../../typings/types";

import style from "./GridItem.css";
import getPlayerSymbol from "../../utils/playerMap";

interface IGridItemProps {
    colN: number;
    rowN: number;
    position: {
        top: boolean;
        bottom: boolean;
        left: boolean;
        right: boolean;
    };
    itemState: IGridItem;
    isFinished: boolean;
    onItemClick: (x: number, y: number) => void;
}

class GridItem extends React.PureComponent<IGridItemProps, any> {
    constructor(props: IGridItemProps) {
        super(props);
        this.onClick = this.onClick.bind(this);
    }

    public render() {
        const {
            position,
            itemState: { player }
        } = this.props;

        return (
            <div
                className={classnames(
                    {
                        [style.finished]: this.props.isFinished
                    },
                    {
                        [style.naught]:
                            player && player.symbol === ISymbol.NAUGHT,
                        [style.cross]: player && player.symbol === ISymbol.CROSS
                    },
                    {
                        [style.root]: true,
                        [style.top]: position.top,
                        [style.bottom]: position.bottom,
                        [style.left]: position.left,
                        [style.right]: position.right
                    }
                )}
                onClick={this.onClick}
            >
                <span>
                    {this.props.itemState.player ? (
                        getPlayerSymbol(this.props.itemState.player.symbol)
                    ) : (
                        <>&nbsp;</>
                    )}
                </span>
            </div>
        );
    }
    private onClick() {
        this.props.onItemClick(this.props.colN, this.props.rowN);
    }
}

export default GridItem;
