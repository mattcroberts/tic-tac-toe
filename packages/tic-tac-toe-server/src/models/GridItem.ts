import Player from "./Player";

class GridItem {
    private player: Player | null = null;

    public placePlayer(player: Player) {
        this.player = player;
    }
}

export default GridItem;
