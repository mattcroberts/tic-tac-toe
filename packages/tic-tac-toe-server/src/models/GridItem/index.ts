import {
    BaseEntity,
    Column,
    Entity,
    PrimaryGeneratedColumn,
    ManyToOne,
    JoinColumn
} from "typeorm";
import Player from "../Player";
import Grid from "../Grid";

export interface IGridItem {
    x: number;
    y: number;
    player: Player | null;
}

@Entity()
export default class GridItem extends BaseEntity implements IGridItem {

    @PrimaryGeneratedColumn("uuid")
    public id!: string;

    @Column({ type: "int" })
    public x: number;

    @Column({ type: "int" })
    public y: number;

    @ManyToOne(type => Grid, grid => grid._gridItems)
    public grid!: Grid;

    @ManyToOne(type => Player, { nullable: true })
    @JoinColumn()
    public player!: Player | null;

    constructor(x: number, y: number) {
        super();
        this.x = x;
        this.y = y;
    }

    public toString() {
return `id=${this.id} xy=${this.x},${this.y} player=${this.player}`
    }
}
