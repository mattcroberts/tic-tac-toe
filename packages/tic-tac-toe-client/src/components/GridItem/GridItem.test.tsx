import * as React from "react";
import { shallow } from "enzyme";
import GridItem from ".";
import { Symbol as ISymbol } from "../../../typings/types";

describe("GridItem", () => {
    const defaultProps = {
        colN: 0,
        rowN: 0,
        position: {
            top: false,
            bottom: false,
            right: false,
            left: false
        },
        itemState: {
            id: "abc123",
            player: null,
            x: 0,
            y: 0
        },
        onItemClick: jest.fn(),
        isFinished: false
    };

    it("should render", () => {
        const component = shallow(<GridItem {...defaultProps} />);
        expect(component).toMatchSnapshot();
    });

    it("should render CROSS", () => {
        const props = {
            ...defaultProps,
            itemState: {
                ...defaultProps.itemState,
                player: { id: "", symbol: ISymbol.CROSS }
            }
        };
        const component = shallow(<GridItem {...props} />);
        expect(component).toMatchSnapshot();
    });

    it("should render NAUGHT", () => {
        const props = {
            ...defaultProps,
            itemState: {
                ...defaultProps.itemState,
                player: { id: "", symbol: ISymbol.NAUGHT }
            }
        };
        const component = shallow(<GridItem {...props} />);
        expect(component).toMatchSnapshot();
    });

    it("should handle item click", () => {
        const props = {
            ...defaultProps,
            itemState: {
                ...defaultProps.itemState,
                player: { id: "", symbol: ISymbol.NAUGHT }
            },
            onItemClick: jest.fn()
        };
        const component = shallow(<GridItem {...props} />);

        component.simulate("click");

        expect(props.onItemClick).toHaveBeenCalled();
        expect(props.onItemClick).toHaveBeenCalledWith(0, 0);
    });
});
