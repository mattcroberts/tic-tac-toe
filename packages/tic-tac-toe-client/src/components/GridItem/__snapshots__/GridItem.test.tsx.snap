// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridItem should render 1`] = `
<div
  className="root"
  onClick={[Function]}
>
  <span>
    Â 
  </span>
</div>
`;

exports[`GridItem should render CROSS 1`] = `
<div
  className="cross root"
  onClick={[Function]}
>
  <span>
    x
  </span>
</div>
`;

exports[`GridItem should render NAUGHT 1`] = `
<div
  className="naught root"
  onClick={[Function]}
>
  <span>
    o
  </span>
</div>
`;
